File: src/xyz/jadonfowler/sbhs/PaletteManager.java
package xyz.jadonfowler.sbhs;

import java.awt.*;
import java.awt.event.*;
import java.awt.image.*;
import java.io.*;
import java.util.*;
import javax.imageio.*;
import javax.swing.*;

public class PaletteManager {
    public static final HashMap<String, String[]> PALETTE_INFO = new HashMap<String, String[]>() {
        private static final long serialVersionUID = 1L;
        {
            String[] g = { "Background (useless)", "Eye color and shoe/glove reflection",
                    "Above the eye, shoe/glove color", "Outside of shoe/glove", "Outline of shoe/glove",
                    "In-ball shine", "Primary fur color", "Secondary fur color", "Third fur color/outline",
                    "Primary skin color", "Secondary skin color", "Third skin color", "Primary shoe color",
                    "Secondary shoe color", "Third shoe color", "Eye Color" };
            put("Sonic", g);
            put("Knuckles", g);
            put("Tails",
                    new String[] { "Background (useless)", "Eye color and shoe/glove reflection",
                            "Above the eye, tail/hand color", "Primary outline", "Secondary outline", "In-ball shine",
                            "Primary fur color", "Secondary fur color", "Third fur color/outline", "Primary skin color",
                            "Secondary skin color", "Third skin color", "Primary shoe color & Weapons",
                            "Secondary shoe color & Weapons", "Third shoe color & Weapons", "Eye Color" });
            put("Amy", new String[] { "Background (useless)", "Eye color and shoe/hammer reflection",
                    "Above the eye, shoe/glove color" });
        }
    };
    
    /**
     * "sonic" -> ["23452345", "23452345", ...]
     */
    public static HashMap<String, String[]> PALETTES = new HashMap<String, String[]>();

    public static void addPaletteTab(JTabbedPane pane, String name, int offset) throws Exception {
        pane.addTab(name, null, createPalettePanel(name, offset), "Edit " + name + " Palette");
    }

    public static JPanel createPalettePanel(String name, int hex) throws Exception {
        String[] colors = new String[16];
        int color = 0;
        String f = "";
        for (int i = hex; i <= hex + 32; i++) {
            SBHS.raf.seek(i);
            int value = SBHS.raf.read();
            // System.out.print(Integer.toHexString(value));
            if (f.length() == 4) {
                f = f.split("(?<=\\G.{2})")[1] + f.split("(?<=\\G.{2})")[0];
                colors[color++] = f;
                f = SBHS.hex(value);
            }
            else {
                f += SBHS.hex(value);
            }
        }
        PALETTES.put(name, colors);
        if (name.equals("Sonic")){
            String[] shadowColors = (String[]) Arrays.asList(colors).toArray();
            shadowColors[0] = GBAColor.toGBA(Color.white);
            shadowColors[1] = GBAColor.toGBA(Color.gray);
            PALETTES.put("Ground Shadow", shadowColors);
        }
        // System.out.println(Arrays.toString(colors));
        int i = 0;
        JPanel jp = new JPanel();
        jp.setLayout(new BoxLayout(jp, BoxLayout.Y_AXIS));
        for (@SuppressWarnings("unused") String s : colors) {
            i++;
            JButton jb = new JButton(name + " Color " + i + ": " + (PALETTE_INFO.get(name) == null ? "Something?"
                    : PALETTE_INFO.get(name).length < i ? "Something?" : PALETTE_INFO.get(name)[i - 1]));
            jb.addActionListener(new ActionListener() {
                final String character = name;

                public void actionPerformed(ActionEvent e) {
                    String name = jb.getText();
                    int i = -1 + (Integer.parseInt(name.split(" ")[2].replace(":", "")));
                    colors[i] = GBAColor.toGBA(SBHS.getColorInput(GBAColor.fromGBA(colors[i])));
                    PALETTES.remove(character);
                    PALETTES.put(character, colors);
                    try {
                        int h1 = Integer.parseInt(colors[i].split("(?<=\\G.{2})")[0], 16);
                        int h2 = Integer.parseInt(colors[i].split("(?<=\\G.{2})")[1], 16);
                        // colors[i] = Integer.toHexString(h2) +
                        // Integer.toHexString(h1) + "";
                        SBHS.raf.seek(hex + (i * 2));
                        SBHS.raf.write(h2);
                        SBHS.raf.seek(hex + (i * 2) + 1);
                        SBHS.raf.write(h1);
                    }
                    catch (IOException e1) {
                        e1.printStackTrace();
                    }
                    // System.out.println(Arrays.toString(colors));
                }
            });
            jp.add(jb);
        }
        JButton upload = new JButton("  Upload Palette");
        upload.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                System.out.println("Uploading palette " + name);
                JFileChooser fc = new JFileChooser();
                if (fc.showOpenDialog(SBHS.frame) == JFileChooser.APPROVE_OPTION) {
                    try {
                        String[] colors = new String[16];
                        BufferedImage img = ImageIO.read(fc.getSelectedFile());
                        for(int x = 0; x < 16; x++) {
                            Color c = new Color(img.getRGB(x, 0));
                            colors[x] = GBAColor.toGBA(c);
                        }
                        PALETTES.remove(name);
                        PALETTES.put(name, colors);
                    }
                    catch (IOException x) {
                        x.printStackTrace();
                    }
                }
            }
        });
        jp.add(upload);
        JButton save = new JButton("  Save Palette");
        save.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                System.out.println("Saving palette " + name);
                //Create Image
                BufferedImage i = new BufferedImage(16, 1, BufferedImage.TYPE_INT_RGB);
                for (int x = 0; x < 16; x++) {
                    Color c = GBAColor.fromGBA(PALETTES.get(name)[x]);
                    i.setRGB(x, 0, c.getRGB());
                }
                //Save Image
                JFileChooser fc = new JFileChooser();
                if (fc.showSaveDialog(SBHS.frame) == JFileChooser.APPROVE_OPTION) {
                    try {
                        File o = fc.getSelectedFile();
                        ImageIO.write(i, "png", o);
                    }
                    catch (IOException x) {
                        x.printStackTrace();
                    }
                }
            }
        });
        jp.add(save);
        return jp;
    }
}

File: src/main/kotlin/xyz/jadonfowler/sbhs/PaletteManager.kt
package xyz.jadonfowler.sbhs

import java.awt.*
import java.awt.event.*
import java.awt.image.*
import java.io.*
import java.util.*
import javax.imageio.*
import javax.swing.*

object PaletteManager {
    val PALETTE_INFO: HashMap<String, Array<String>> = object : HashMap<String, Array<String>>() {
        init {
            val g = arrayOf("Background (useless)", "Eye color and shoe/glove reflection", "Above the eye, shoe/glove color", "Outside of shoe/glove", "Outline of shoe/glove", "In-ball shine", "Primary fur color", "Secondary fur color", "Third fur color/outline", "Primary skin color", "Secondary skin color", "Third skin color", "Primary shoe color", "Secondary shoe color", "Third shoe color", "Eye Color")
            put("Sonic", g)
            put("Knuckles", g)
            put("Tails",
                    arrayOf("Background (useless)", "Eye color and shoe/glove reflection", "Above the eye, tail/hand color", "Primary outline", "Secondary outline", "In-ball shine", "Primary fur color", "Secondary fur color", "Third fur color/outline", "Primary skin color", "Secondary skin color", "Third skin color", "Primary shoe color & Weapons", "Secondary shoe color & Weapons", "Third shoe color & Weapons", "Eye Color"))
            put("Amy", arrayOf("Background (useless)", "Eye color and shoe/hammer reflection", "Above the eye, shoe/glove color"))
        }
    }

    /**
     * "sonic" -> ["23452345", "23452345", ...]
     */
    var PALETTES = HashMap<String, Array<String>>()
    var PALETTE_OFFSET = HashMap<String, Int>()

    @Throws(Exception::class)
    fun addPaletteTab(pane: JTabbedPane, name: String, offset: Int) {
        pane.addTab(name, null, createPalettePanel(name, offset), "Edit $name Palette")
    }

    @Throws(Exception::class)
    fun createPalettePanel(name: String, hex: Int): JPanel {
        PALETTE_OFFSET.put(name, hex)
        val colors = arrayOfNulls<String>(16)
        var color = 0
        var f = ""
        for (i in hex..hex + 32) {
            SBHS.raf.seek(i.toLong())
            val value = SBHS.raf.read()
            //System.out.println(value + " " + f);
            // System.out.print(Integer.toHexString(value));
            if (f.length == 4) {
                f = f.split("(?<=\\G.{2})".toRegex()).dropLastWhile(String::isEmpty).toTypedArray()[1] +
                        f.split("(?<=\\G.{2})".toRegex()).dropLastWhile(String::isEmpty).toTypedArray()[0]
                colors[color++] = f
                //System.out.println("@"+i + " " + color + " "+f);
                f = SBHS.hex(value)
            } else {
                f += SBHS.hex(value)
            }
        }
        PALETTES.put(name, colors.map { it.orEmpty() }.toTypedArray())
        //SpriteManager.printPalette(PALETTES.get(name));
        if (name == "Sonic") {
            val shadowColors = Arrays.asList<String>(*colors).toTypedArray()
            shadowColors[0] = GBAColor.toGBA(Color.white)
            shadowColors[1] = GBAColor.toGBA(Color.gray)
            PALETTES.put("Ground Shadow", shadowColors)
        }
        // System.out.println(Arrays.toString(colors));
        var i = 0
        val jp = JPanel()
        jp.layout = BoxLayout(jp, BoxLayout.Y_AXIS)
        for (s in colors) {
            i++
            val jb = JButton("$name Color $i: " + if (PALETTE_INFO[name] == null)
                "Something?"
            else if (PALETTE_INFO[name]!!.size < i) "Something?" else PALETTE_INFO[name]!![i - 1])
            jb.addActionListener(object : ActionListener {
                internal val character = name

                override fun actionPerformed(e: ActionEvent) {
                    val name = jb.text
                    val i = -1 + Integer.parseInt(name.split(" ".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()[2].replace(":", ""))
                    colors[i] = GBAColor.toGBA(SBHS.getColorInput(GBAColor.fromGBA(colors[i].orEmpty())))
                    PALETTES.remove(character)
                    PALETTES.put(character, colors.map { it.orEmpty() }.toTypedArray())
                    try {
                        val h1 = Integer.parseInt(colors[i]?.split("(?<=\\G.{2})".toRegex())?.dropLastWhile({ it.isEmpty() })?.toTypedArray()?.get(0), 16)
                        val h2 = Integer.parseInt(colors[i]?.split("(?<=\\G.{2})".toRegex())?.dropLastWhile({ it.isEmpty() })?.toTypedArray()?.get(1), 16)
                        // colors[i] = Integer.toHexString(h2) +
                        // Integer.toHexString(h1) + "";
                        SBHS.raf.seek((hex + i * 2).toLong())
                        SBHS.raf.write(h2)
                        SBHS.raf.seek((hex + i * 2 + 1).toLong())
                        SBHS.raf.write(h1)
                    } catch (e1: IOException) {
                        e1.printStackTrace()
                    }

                    // System.out.println(Arrays.toString(colors));
                }
            })
            jp.add(jb)
        }
        val upload = JButton("  Upload Palette")
        upload.addActionListener {
            println("Uploading palette " + name)
            val fc = JFileChooser()
            if (fc.showOpenDialog(SBHS.frame) == JFileChooser.APPROVE_OPTION) {
                try {
                    val img = ImageIO.read(fc.selectedFile)
                    for (x in 0..15) {
                        val c = Color(img.getRGB(x, 0))
                        println("Upload-color: " + GBAColor.toGBA(c))
                        colors[x] = GBAColor.toGBA(c)
                        try {
                            val h1 = Integer.parseInt(colors[x]?.split("(?<=\\G.{2})".toRegex())?.dropLastWhile(String::isEmpty)?.toTypedArray()?.get(0), 16)
                            val h2 = Integer.parseInt(colors[x]?.split("(?<=\\G.{2})".toRegex())?.dropLastWhile(String::isEmpty)?.toTypedArray()?.get(1), 16)
                            SBHS.raf.seek((hex + x * 2).toLong())
                            SBHS.raf.write(h2)
                            SBHS.raf.seek((hex + x * 2 + 1).toLong())
                            SBHS.raf.write(h1)
                        } catch (e1: IOException) {
                            e1.printStackTrace()
                        }

                    }
                    SpriteManager.printPalette(PALETTES[name]!!)
                    PALETTES.remove(name)
                    PALETTES.put(name, colors.map { it.orEmpty() }.toTypedArray())
                    SpriteManager.printPalette(PALETTES[name]!!)
                } catch (x: IOException) {
                    x.printStackTrace()
                }

            }
        }
        jp.add(upload)
        val save = JButton("  Save Palette")
        save.addActionListener {
            println("Saving palette " + name)
            //Create Image
            val i = BufferedImage(16, 1, BufferedImage.TYPE_INT_RGB)
            for (x in 0..15) {
                val c = GBAColor.fromGBA(PALETTES[name]!![x])
                i.setRGB(x, 0, c.rgb)
            }
            //Save Image
            val fc = JFileChooser()
            if (fc.showSaveDialog(SBHS.frame) == JFileChooser.APPROVE_OPTION) {
                try {
                    val o = fc.selectedFile
                    ImageIO.write(i, "png", o)
                } catch (x: IOException) {
                    x.printStackTrace()
                }

            }
        }
        jp.add(save)
        return jp
    }
}
